sslr-core/src/main/java/com/sonar/sslr/impl/typed/SyntaxTreeCreator.java

```diff
@@ -0,0 +1,193 @@
+import com.sonar.sslr.api.typed.ActionParser.GrammarBuilderInterceptor;
+import com.sonar.sslr.api.typed.Input;
+import com.sonar.sslr.api.typed.NodeBuilder;
+import com.sonar.sslr.api.typed.Optional;
+import org.sonar.sslr.grammar.GrammarRuleKey;
+import org.sonar.sslr.internal.grammar.MutableParsingRule;
+import org.sonar.sslr.internal.matchers.ParseNode;
+import org.sonar.sslr.internal.vm.TokenExpression;
+import org.sonar.sslr.internal.vm.TriviaExpression;
+
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+public class SyntaxTreeCreator<T> {
+
+  private final Object treeFactory;
+  private final GrammarBuilderInterceptor mapping;
+  private final NodeBuilder nodeBuilder;
+
+  private final Token.Builder tokenBuilder = Token.builder();
+  private final List<Trivia> trivias = new ArrayList<>();
+
+  private Input input;
+
+  public SyntaxTreeCreator(Object treeFactory, GrammarBuilderInterceptor mapping, NodeBuilder nodeBuilder) {
+    this.treeFactory = treeFactory;
+    this.mapping = mapping;
+    this.nodeBuilder = nodeBuilder;
+  }
+
+  public T create(ParseNode node, Input input) {
+    this.input = input;
+    this.trivias.clear();
+    T result = (T) visit(node);
+    if (result instanceof AstNode) {
+      ((AstNode) result).hasToBeSkippedFromAst();
+    }
+    return result;
+  }
+
+  private Object visit(ParseNode node) {
+    if (node.getMatcher() instanceof MutableParsingRule) {
+      return visitNonTerminal(node);
+    } else {
+      return visitTerminal(node);
+    }
+  }
+
+  private Object visitNonTerminal(ParseNode node) {
+    MutableParsingRule rule = (MutableParsingRule) node.getMatcher();
+    GrammarRuleKey ruleKey = rule.getRuleKey();
+    Method method = mapping.actionForRuleKey(ruleKey);
+
+    Object result;
+
+    if (mapping.hasMethodForRuleKey(ruleKey)) {
+
+      // TODO Drop useless intermediate nodes
+      Preconditions.checkState(node.getChildren().size() == 1);
+      result = visit(node.getChildren().get(0));
+
+    } else if (mapping.isOptionalRule(ruleKey)) {
+
+      Preconditions.checkState(node.getChildren().size() <= 1);
+      if (node.getChildren().isEmpty()) {
+        result = Optional.absent();
+      } else {
+        Object child = visit(node.getChildren().get(0));
+        if (child instanceof AstNode) {
+          ((AstNode) child).hasToBeSkippedFromAst();
+        }
+        result = Optional.of(child);
+      }
+
+    } else if (mapping.isOneOrMoreRule(ruleKey)) {
+
+      result = convertChildren(node);
+
+    } else if (mapping.isZeroOrMoreRule(ruleKey)) {
+
+      List<Object> convertedChildren = convertChildren(node);
+      result = convertedChildren.isEmpty() ? Optional.absent() : Optional.of(convertedChildren);
+
+    } else if (method == null) {
+
+      result = nodeBuilder.createNonTerminal(rule.getRuleKey(), rule, convertChildren(node), node.getStartIndex(), node.getEndIndex());
```

Comment:
so this is being called for the Java grammar for all rules defined in the `JavaLexer` class right? (those don't have an associated method) 

it sounds a bit confusing... I think in the long run we will want to drop this `createNonTerminal()` method